dali/operators/reader/loader/video/video_file.cc

```diff
@@ -0,0 +1,232 @@
+// Copyright (c) 2021, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/reader/loader/video/video_file.h"
+#include "dali/core/error_handling.h"
+
+namespace dali {
+
+void VideoFileCPU::InitAvState() {
+  codec_ctx_ = avcodec_alloc_context3(codec_);
+  DALI_ENFORCE(codec_ctx_, "Could not create av codec context");
+
+  DALI_ENFORCE(
+    avcodec_parameters_to_context(codec_ctx_, codec_params_) >= 0,
+    "Could not fill the codec based on parameters");
+
+  DALI_ENFORCE(
+    avcodec_open2(codec_ctx_, codec_, nullptr) == 0,
+    "Could not initialize codec context");
+
+  frame_ = av_frame_alloc();
+  DALI_ENFORCE(frame_, "Could not allocate the av frame");
+
+  packet_ = av_packet_alloc();
+  DALI_ENFORCE(packet_, "Could not allocate av packet");
+}
+
+void VideoFileCPU::FindVideoStream() {
+  for (size_t i = 0; i < ctx_->nb_streams; ++i) {
+    auto stream = ctx_->streams[i];
+    codec_params_ = ctx_->streams[i]->codecpar;
+    codec_ = avcodec_find_decoder(codec_params_->codec_id);
+
+    if (codec_ == nullptr) {
+      continue;
+    }
+
+    if (codec_->type == AVMEDIA_TYPE_VIDEO) {
+        stream_id_ = i;
+        return;
+    }
+  }
+
+  DALI_FAIL(make_string("Could not find a valid video stream in file ", filename_));
+}
+
+VideoFileCPU::VideoFileCPU(std::string &filename) : 
+  filename_(filename) {
+  ctx_ = avformat_alloc_context();
+  DALI_ENFORCE(ctx_, "Could not create avformat context");
+
+  DALI_ENFORCE(
+    avformat_open_input(&ctx_, filename.c_str(), nullptr, nullptr) == 0,
+    make_string("Failed to open video file at path ", filename));
+
+  FindVideoStream();
+  InitAvState();
+  BuildIndex();
+}
+
+void VideoFileCPU::BuildIndex() {
+  // TODO(awolant): Optimize this function for: 
+  //  - CFR
+  //  - index present in the header
+
+  int last_keyframe = -1;
+  while (ReadRegularFrame(nullptr, false)) {
```

Comment:
So we need to decode the video to build the index, right?

Reply:
Nvm. It just parses it.

Reply:
It does. In general scenario, when we know nothing about the video, this is the only way, I think. I made it as thin as I could. There are various cases, where we can optimize the index building. It will happen in consecutive PRs, as for know I wanted something that works.